algorithms/kernel/dtrees/forest/regression/oneapi/cl_kernels/df_batch_regression_kernels.cl

```diff
@@ -0,0 +1,994 @@
+
+                nodeR[0] = nodeL[0] + nodeL[1];
+                nodeR[1] = nodeP[1] - nodeL[1];
+                nodeR[2] = badVal;
+                nodeR[3] = badVal;
+                nodeR[4] = nodeR[1]; // num of items in Left part = nRows in new node
+            }
+            nCreatedNodes += sub_group_reduce_add(splitNode) * 2;
+        }
+    }
+
+    __kernel void splitNodeListOnGroupsBySize(const __global int * nodeList, int nNodes, __global int * bigNodesGroups, __global int * nodeIndeces,
+                                              int minRowsBlock) {
+        /*for now only 3 groups are produced, may be more required*/
+        const int bigNodeLowBorderBlocksNum = 32; // fine, need to experiment and find better one
+        const int blockSize                 = minRowsBlock;
+        const int nNodeProp                 = 5; // num of split attributes for node
+
+        const int local_id   = get_sub_group_local_id();
+        const int nodeId     = get_global_id(1);
+        const int local_size = get_sub_group_size();
+
+        int nBigNodes       = 0;
+        int maxBigBlocksNum = 1;
+        int nMidNodes       = 0;
+        int maxMidBlocksNum = 1;
+
+        /*calculate num of big and mid nodes*/
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            int nRows   = nodeList[i * nNodeProp + 1];
+            int nBlocks = nRows / blockSize + !!(nRows % blockSize);
+            int bigNode = 0;
+            int midNode = 0;
+            if (nBlocks > bigNodeLowBorderBlocksNum)
+                bigNode = 1;
+            else if (nBlocks > 1)
+                midNode = 1;
+
+            nBigNodes += sub_group_reduce_add(bigNode);
+            nMidNodes += sub_group_reduce_add(midNode);
+            maxBigBlocksNum = max(maxBigBlocksNum, bigNode ? nBlocks : 0);
+            maxBigBlocksNum = sub_group_reduce_max(maxBigBlocksNum);
+            maxMidBlocksNum = max(maxMidBlocksNum, midNode ? nBlocks : 0);
+            maxMidBlocksNum = sub_group_reduce_max(maxMidBlocksNum);
+        }
+
+        nBigNodes = sub_group_broadcast(nBigNodes, 0);
+        nMidNodes = sub_group_broadcast(nMidNodes, 0);
+
+        if (0 == local_id)
+        {
+            bigNodesGroups[0] = nBigNodes;
+            bigNodesGroups[1] = maxBigBlocksNum;
+            bigNodesGroups[2] = nMidNodes;
+            bigNodesGroups[3] = maxMidBlocksNum;
+            bigNodesGroups[4] = nNodes - nBigNodes - nMidNodes;
+            bigNodesGroups[5] = 1;
+        }
+
+        int sumBig = 0;
+        int sumMid = 0;
+
+        /*split nodes on groups*/
+        for (int i = local_id; i < nNodes; i += local_size)
+        {
+            int nRows   = nodeList[i * nNodeProp + 1];
+            int nBlocks = nRows / blockSize + !!(nRows % blockSize);
+            int bigNode = 0;
+            int midNode = 0;
+            if (nBlocks > bigNodeLowBorderBlocksNum)
```

Comment:
It is probably better to initialize this variables using bool to int casts:
`int binNode = (int)(nBlocks > bigNodeLowBorderBlockSum)`
`int midNode = (int)(nBlocks <= binNodeLowBorderBlockSum && nBlocks > 1)`

Reply:
good, thank you!

Reply:
applied